src/helpers/validators/validateString/index.ts

```diff
@@ -0,0 +1,128 @@
+const UPPERCASE_MIN = 97;
+const UPPERCASE_MAX = 122;
+
+// define minimum and maximum range for ASCII digits
+const DIGIT_MIN = 48;
+const DIGIT_MAX = 57;
+
+/**
+ * Determines whether a given input string is considered valid based on
+ *   the provided requirements.
+ * @author Austen Money
+ * @param input input string
+ * @param reqs requirements that the provided string should conform to
+ * @returns whether input is considered valid according to reqs - if
+ *   valid, returns a cleaned version of input; if invalid, returns
+ *   a string containing error messages describing which requirements
+ *   were not met.
+ */
+const validateString = (
+  input: string,
+  reqs: StringValidationRequirements,
+): ValidationResult => {
+  // stores all invalid input errors
+  const errorMessages: string[] = [];
+
+  // contains version of input that will be returned
+  let cleanedValue: string = input;
+
+  // remove whitespace if required
+  if (reqs.ignoreWhitespace) {
+    cleanedValue = input.replace(/\s+/g, '');
+  }
+
+  // apply max char requirement
+  if (reqs.minLen) {
+    if (cleanedValue.length < reqs.minLen) {
+      errorMessages.push(`Input must not be under ${reqs.minLen} character(s)`);
+    }
+  }
+
+  // apply max char requirement
+  if (reqs.maxLen) {
+    if (cleanedValue.length > reqs.maxLen) {
+      errorMessages.push(`Input must not be over ${reqs.maxLen} character(s)`);
+    }
+  }
+
+  // apply alphabetical requirement
+  if (reqs.lettersOnly) {
+    // remove diacritics
+    cleanedValue = cleanedValue.normalize('NFD').replace(/[\u0300-\u036f]/g, '');
+
+    for (let i = 0, len = cleanedValue.length; i < len; i++) {
+      const curr = cleanedValue.charCodeAt(i);
+
+      // check that char is an upper or lower-case letter
+      if (!(curr >= LOWERCASE_MIN && curr <= LOWERCASE_MAX)
+        && !(curr >= UPPERCASE_MIN && curr <= UPPERCASE_MAX)) {
+        errorMessages.push('Input must only contain letters');
+        break;
+      }
+    }
+  }
+
+  // apply numerical requirement
+  if (reqs.numbersOnly) {
+    for (let i = 0, len = cleanedValue.length; i < len; i++) {
+      const curr = cleanedValue.charCodeAt(i);
+
+      // check that char is a digit
+      if (!(curr >= DIGIT_MIN && curr <= DIGIT_MAX)) {
+        errorMessages.push('Input must only contain numbers');
+        break;
+      }
+    }
+  }
+
+  // apply regex requirement
+  if (reqs.regexTest) {
+    const regex = reqs.regexTest;
+
+    // validate and create customized error message if description is provided
+    const result = validateRegex({
+      input: cleanedValue,
+      regex,
+      regexDescription: reqs.regexDescription,
+    });
+
+    // if string did not pass regex validation, add error message
+    if (result.isValid === false) {
+      errorMessages.push(result.errorMessage);
+    }
+  }
+
+  // combine all error messages into one string to return
+  const errorMessage = `The following error(s) occurred: ${errorMessages.join(', ')}.`;
```

Comment:
This gets me thinking that we could do with a `genCommaList` function that turns a list of strings into an Oxford comma list:

```ts
genCommaList(['apple']) // > 'apple'
genCommaList(['apple', 'orange']) // > 'apple and orange'
genCommaList(['apple', 'orange', 'mango']) // > 'apple, orange, and mango'
```

It seems like this would be a function that would be useful here but also in many other projects. What do you think?

Reply:
I like that! It does seem like an issue that would crop up in a lot of other situations.

Reply:
Yay thank you!